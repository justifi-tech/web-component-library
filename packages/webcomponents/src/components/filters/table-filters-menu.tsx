import { Component, Listen, Prop, State, Watch, h } from '@stencil/core';
import { Button } from '../../ui-components/button';
import { hasFilters } from './utils';
import { dropdownMenu } from '../../styles/parts';

@Component({
  tag: 'table-filters-menu'
})
export class TableFiltersMenu {
  @State() anchorButton: HTMLElement;
  @State() menu: HTMLElement;
  @State() showClearButton: boolean = false;

  @Prop() params: any;
  @Prop() clearParams: () => void;

  @Watch('params')
  watchParamsHandler(newValue: any) {
    this.showClearButton = hasFilters(newValue);
  }

  @Listen('showEvent')
  show() {
    this.menu.classList.add('show');
  }

  @Listen('hideEvent')
  hide() {
    this.menu.classList.remove('show');
  }

  componentWillLoad() {
    this.showClearButton = hasFilters(this.params);
  }

  render() {
    return (
      <div class="d-flex gap-2">
        <div class="dropdown pb-3">
          <Button
            variant="primary"
            type="button dropdown-toggle"
            ref={(el) => (this.anchorButton = el)}
            class="btn btn-primary dropdown-toggle"
            data-test-id="open-filters-button"
            data-bs-toggle="dropdown"
          >
            {"Filters"}
          </Button>
          <custom-popper
            offset={[20, 10]}
            strategy="fixed"
            anchorRef={this.anchorButton}
            trigger="click"
          >
            <form
              class="dropdown-menu"
              part={dropdownMenu}
              data-test-id="filter-menu"
              ref={(el) => (this.menu = el)}
              onSubmit={(e) => e.preventDefault()}
            >
              {/* Render filter inputs here */}
              <slot />
            </form>
          </custom-popper>
        </div>
        <div>
          <Button
            variant="secondary"
            onClick={() => this.clearParams()}
            hidden={!this.showClearButton}
            data-test-id="clear-filters-button"
          >
            {"Clear Filters"}
          </Button>
        </div>
      </div>
    );
  }
}
