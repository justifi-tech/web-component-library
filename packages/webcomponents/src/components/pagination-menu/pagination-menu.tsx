import { Component, h, Prop } from '@stencil/core';
import { ExtendedPagingDefaults, ExtendedPagingInfo } from '../../api/Pagination';
import { StyledHost } from '../../ui-components';

export interface PaginationProps {
  paging: ExtendedPagingInfo;
  params?: {};
}

@Component({
  tag: 'pagination-menu'
})

export class PaginationMenu {
  @Prop() paging: PaginationProps['paging'] = ExtendedPagingDefaults;
  @Prop() params: PaginationProps['params'] = {};

  render() {
    return (
      <StyledHost exportparts="pagination-buttons-container">
        <nav
          aria-label="Table pagination"
          class="d-flex justify-content-end gap-3"
          part="pagination-bar"
        >
          <button
            class="page-link"
            onClick={(e) => {
              e.preventDefault();
              this.paging.handleClickPrevious(this.paging.start_cursor);
            }}
            part={`button-link ${this.paging.has_next ? '' : ' button-disabled'}`}
          >
            <span class="me-1" part="page-arrow">&laquo;</span>
            <span part="page-button-text">Previous</span>
          </button>
          <button
            class="page-link"
            onClick={(e) => {
              e.preventDefault();
              this.paging.handleClickNext(this.paging.end_cursor);
            }}
            part={`button-link ${this.paging.has_next ? '' : ' button-disabled'}`}
          >
            <span part="page-button-text">Next</span>
            <span class="ms-1" part="page-arrow">&raquo;</span>
          </button>
        </nav>
      </StyledHost >
    );
  }
}
