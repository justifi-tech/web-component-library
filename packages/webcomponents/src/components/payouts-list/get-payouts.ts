import { Payout } from '../../api';
import { getErrorMessage } from '../../api/services/utils';

export const makeGetPayouts =
  ({ id, authToken, service }) =>
  async ({ params, onSuccess, onError }) => {
    try {
      const response = await service.fetchPayouts(id, authToken, params);

      if (!response.error) {
        const pagingInfo = {
          ...response.page_info,
        };

        const payouts =
          response.data?.map((dataItem) => new Payout(dataItem)) || [];

        onSuccess({ payouts, pagingInfo });
      } else {
        const responseError = getErrorMessage(response.error);
        return onError(responseError);
      }
    } catch (error) {
      return onError(error.message || error);
    }
  };
