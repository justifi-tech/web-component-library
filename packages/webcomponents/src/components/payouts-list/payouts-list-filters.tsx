import { Component, h, Prop } from '@stencil/core';
import { filterParams, propsParams, clearParams } from './payouts-list-params-state';
import { StyledHost } from '../../ui-components';
import { convertToLocal, convertToUTC } from '../../utils/utils';
import { createdAfterPayoutsListFilterParam, payoutsListFilterMenu } from '../../styles/parts';
@Component({
  tag: 'justifi-payouts-list-filters',
  shadow: true
})
export class PayoutsListFilters {
  @Prop() createdAfter?: string;
  @Prop() createdBefore?: string;

  componentWillLoad() {
    const propsToSet = {
      created_after: this.createdAfter,
      created_before: this.createdBefore
    };

    Object.entries(propsToSet).forEach(([key, value]) => {
      if (value) {
        propsParams[key] = value;
      }
    });
  }
  
  setParamsOnChange = (name: string, value: string) => {
    filterParams[name] = value;
  }

  handleDateInput = (name: string, value: string) => {
    const utcDate = convertToUTC(value, { setEndOfDay: true });
    this.setParamsOnChange(name, utcDate);
  }

  render() {
    const filterMenuParams = { ...filterParams }

    return (
      <StyledHost>
        <table-filters-menu params={filterMenuParams} clearParams={clearParams} part={payoutsListFilterMenu}>
          <div class="grid-cols-2 gap-3 p-1">
            <div class="p-2">
              <form-control-date
                name="created_after"
                label="Created After"
                inputHandler={this.handleDateInput}
                defaultValue={
                  convertToLocal(this.createdAfter, { showInputDate: true }) ||
                  convertToLocal(filterParams.created_after, { showInputDate: true })
                }
                disabled={!!this.createdAfter}
                part={createdAfterPayoutsListFilterParam}
              />
            </div>
            <div class="p-2">
              <form-control-date
                name="created_before"
                label="Created Before"
                inputHandler={this.handleDateInput}
                defaultValue={
                  convertToLocal(this.createdBefore, { showInputDate: true }) ||
                  convertToLocal(filterParams.created_before, { showInputDate: true })
                }
                disabled={!!this.createdBefore}
                part={createdAfterPayoutsListFilterParam}
              />
            </div>
          </div>
        </table-filters-menu>
      </StyledHost>
    );
  }
}
