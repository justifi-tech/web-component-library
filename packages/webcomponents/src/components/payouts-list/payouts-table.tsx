import { h } from '@stencil/core';
import { convertToLocal } from '../../utils/utils';
import { MapPayoutStatusToBadge } from './payouts-status';
import { getAlternateTableCellPart, tableHeadCell } from '../../styles/parts';
import { Payout } from '../../api/Payout';

export const defaultColumnsKeys = 'created_at,amount,status,payments_total,refunds_total,fees_total,other_total,csv';

export const payoutTableColumns = {
  created_at: () => (
    <th part={tableHeadCell} scope="col" title="The date the payout was deposited">
      Created
    </th>
  ),
  sub_account_name: () => (
    <th part={tableHeadCell} scope="col" title="The sub account associated with the payout">
      Sub Account
    </th>
  ),
  payments_total: () => (
    <th part={tableHeadCell} scope="col" title="The total payments amount associated with the payout">
      Payments
    </th>
  ),
  refunds_total: () => (
    <th part={tableHeadCell} scope="col" title="The total refunds amount associated with the payout">
      Refunds
    </th>
  ),
  fees_total: () => (
    <th part={tableHeadCell} scope="col" title="The total fees amount associated with the payout">
      Fees
    </th>
  ),
  other_total: () => (
    <th part={tableHeadCell} scope="col" title="The total other amount">
      Other
    </th>
  ),
  amount: () => (
    <th part={tableHeadCell} scope="col" title="The total amount of the payout">
      Amount
    </th>
  ),
  status: () => (
    <th part={tableHeadCell} scope="col" title="The real-time status of each payout">
      Status
    </th>
  ),
  csv: () => (
    <th part={tableHeadCell} scope="col" title="Export CSV" />
  ),
};

export const payoutTableCells = (downloadCSV) => ({
  created_at: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)}>
      <div class="fw-bold">{convertToLocal(payout.created_at, { showDisplayDate: true })}</div>
    </td>
  ),
  sub_account_name: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)}>{payout.sub_account_name}</td>
  ),
  payments_total: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)}>{payout.formatPaymentAmount(payout.payments_total, true)}</td>
  ),
  refunds_total: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)}>{payout.formatPaymentAmount(payout.refunds_total, true)}</td>
  ),
  fees_total: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)}>{payout.formatPaymentAmount(payout.fees_total, true)}</td>
  ),
  other_total: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)}>{payout.formatPaymentAmount(payout.other_total, true)}</td>
  ),
  amount: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)}>{payout.formatPaymentAmount(payout.amount, true)}</td>
  ),
  status: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)} innerHTML={MapPayoutStatusToBadge(payout.status)}></td>
  ),
  csv: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)}>
      <a href="#" onClick={(event) => { event.preventDefault(); downloadCSV(payout); }}>
        CSV
      </a>
    </td>
  ),
});
