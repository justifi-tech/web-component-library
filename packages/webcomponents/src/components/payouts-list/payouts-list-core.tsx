import { Component, h, Prop, State, Watch, Event, EventEmitter } from '@stencil/core';
import { PagingInfo, Payout, SubAccount, pagingDefaults } from '../../api';
import { ComponentError, ComponentErrorCodes, ComponentErrorSeverity } from '../../api/ComponentError';
import { TableEmptyState, TableErrorState, TableLoadingState } from '../../ui-components';
import { payoutTableCells, payoutTableColumns } from './payouts-table';
import { Table } from '../../utils/table';
import { queryParams, onQueryParamsChange } from './payouts-list-params-state';

@Component({
  tag: 'payouts-list-core',
})

export class PayoutsListCore {
  @Prop() getPayouts: Function;
  @Prop() getSubAccounts: Function;
  @Prop() getPayoutCSV: Function;
  @Prop() columns: string;

  @State() payouts: Payout[] = [];
  @State() payoutsTable: Table;
  @State() subAccounts: SubAccount[] = [];
  @State() loading: boolean = true;
  @State() errorMessage: string;
  @State() paging: PagingInfo = pagingDefaults;
  @State() pagingParams: any = {};

  @Watch('pagingParams')
  @Watch('getPayouts')
  @Watch('getSubAccounts')
  updateOnPropChange() {
    this.fetchData();
  }

  @Event({
    eventName: 'row-clicked',
    bubbles: true,
  }) rowClicked: EventEmitter<Payout>;

  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentError>;

  componentWillLoad() {
    this.payoutsTable = new Table(this.payouts, this.columns, payoutTableColumns, payoutTableCells(this.downloadCSV));
    if (this.getPayouts && this.getSubAccounts) {
      this.fetchData();
    }
 
    onQueryParamsChange('set', () => {
      delete this.pagingParams.before_cursor;
      delete this.pagingParams.after_cursor;
      this.fetchData();
    });

    onQueryParamsChange('reset', () => {
      this.pagingParams = {};
    });
  }

  fetchData(): void {
    this.loading = true;

    this.getPayouts({
      params: this.requestParams,
      onSuccess: ({ payouts, pagingInfo }) => {
        this.payouts = payouts;
        this.paging = pagingInfo;
        this.payoutsTable.collectionData = this.payouts;
        const shouldFetchSubAccounts = this.payoutsTable.columnKeys.includes('sub_account_name');

        if (shouldFetchSubAccounts) {
          this.fetchSubAccounts();
        } else {
          this.loading = false;
        }

      },
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.errorEvent.emit({
          errorCode: code,
          message: error,
          severity,
        });
        this.loading = false;
      },
    });
  }

  async fetchSubAccounts(): Promise<void> {
    this.getSubAccounts({
      params: this.subAccountParams,
      onSuccess: ({ subAccounts }) => {
        this.subAccounts = subAccounts;
        this.payouts = this.payouts.map((payout) => {
          payout.sub_account_name = this.subAccounts.find((subAccount) => subAccount.id === payout.account_id)?.name;
          return payout;
        });
        this.loading = false;
      },
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.errorEvent.emit({
          errorCode: code,
          message: error,
          severity,
        });
        this.loading = false;
      },
    });
  }

  downloadCSV = (payoutId: string) => {
    this.getPayoutCSV({
      payoutId,
      onError: () => {
        this.errorEvent.emit({
          errorCode: ComponentErrorCodes.FETCH_ERROR,
          message: 'Failed to download CSV',
          severity: ComponentErrorSeverity.ERROR,
        });
      },
    });
  }

  handleClickPrevious = (beforeCursor: string) => {
    this.pagingParams = { before_cursor: beforeCursor };
  };

  handleClickNext = (afterCursor: string) => {
    this.pagingParams = { after_cursor: afterCursor };
  };

  rowClickHandler = (e) => {
    const clickedPayoutID = e.target.closest('tr').dataset.rowEntityId;
    if (!clickedPayoutID) return;
    this.rowClicked.emit(this.payouts.find((payout) => payout.id === clickedPayoutID));
  }

  get requestParams() {
    const combinedParams = { ...queryParams, ...this.pagingParams };
    return combinedParams;
  }

  get subAccountParams() {
    let accountIdNumbers = this.payouts.map((payout) => payout.account_id);
    let uniqueAccountIds = [...new Set(accountIdNumbers)];
    let accountIdString = uniqueAccountIds.join(',');
    return { sub_account_id: accountIdString };
  }

  get entityId() {
    return this.payouts.map((payout) => payout.id);
  }

  get showEmptyState() {
    return !this.loading && !this.errorMessage && this.payoutsTable.rowData.length < 1;
  }

  get showErrorState() {
    return !this.loading && !!this.errorMessage;
  }

  get showRowData() {
    return !this.showEmptyState && !this.showErrorState && !this.loading;
  }

  render() {
    return (
      <div>
        <div class="table-wrapper" part="table-wrapper">
          <table class="table table-hover" part="table">
            <thead class="table-head sticky-top" part="table-head">
              <tr class="table-light text-nowrap" part="table-head-row">
                {this.payoutsTable.columnData.map((column) => column)}
              </tr>
            </thead>
            <tbody class="table-body" part="table-body">
              <TableLoadingState
                  columnSpan={this.payoutsTable.columnKeys.length}
                  isLoading={this.loading}
                />
                <TableEmptyState
                  isEmpty={this.showEmptyState}
                  columnSpan={this.payoutsTable.columnKeys.length}
                />
                <TableErrorState
                  columnSpan={this.payoutsTable.columnKeys.length}
                  errorMessage={this.errorMessage}
                />
                {this.showRowData && this.payoutsTable.rowData.map((data, index) => (
                  <tr
                    data-test-id="table-row"
                    data-row-entity-id={this.entityId[index]}
                    onClick={this.rowClickHandler}
                    part={`table-row ${index % 2 ? "table-row-even" : "table-row-odd"}`}
                  >
                    {data}
                  </tr>
                ))}
            </tbody>
            {this.paging && (
              <tfoot class="sticky-bottom">
                <tr class="table-light align-middle">
                  <td part="pagination-bar" colSpan={this.payoutsTable.columnData.length}>
                    <pagination-menu
                      paging={{
                        ...this.paging,
                        handleClickPrevious: this.handleClickPrevious,
                        handleClickNext: this.handleClickNext,
                      }}
                    />
                  </td>
                </tr>
              </tfoot>
            )}
          </table>
        </div>
      </div>
    );
  }
}
