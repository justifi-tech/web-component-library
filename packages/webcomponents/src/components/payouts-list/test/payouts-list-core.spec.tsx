jest.mock('../../../ui-components/styled-host/styled-host.css', () => '');

import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { PaginationMenu } from '../../pagination-menu/pagination-menu';
import { PayoutsListCore } from '../payouts-list-core';
import mockSuccessResponse from '../../../../../../mockData/mockPayoutsSuccess.json';
import mockSubAccountSuccessResponse from '../../../../../../mockData/mockSubAccountsListSuccess.json';
import { IApiResponseCollection, IPayout, ISubAccount } from '../../../api';
import { makeGetPayouts } from '../get-payouts';
import { makeGetSubAccounts } from '../../../api/get-subaccounts';
import { defaultColumnsKeys } from '../payouts-table';
import { TableFiltersMenu } from '../../../ui-components/filters/table-filters-menu';
import { PayoutsListFilters } from '../payouts-list-filters';

const mockPayoutsResponse = mockSuccessResponse as IApiResponseCollection<IPayout[]>;
const mockSubAccountsResponse = mockSubAccountSuccessResponse as IApiResponseCollection<ISubAccount[]>;
const components = [PayoutsListCore, PaginationMenu, TableFiltersMenu, PayoutsListFilters];

describe('payouts-list-core', () => {
  it('renders properly', async () => {
    const mockPayoutService = {
      fetchPayouts: jest.fn().mockResolvedValue(mockPayoutsResponse),
    };
    const getPayouts = makeGetPayouts({
      id: '123',
      authToken: '123',
      service: mockPayoutService
    });

    const mockSubAccountsService = {
      fetchSubAccounts: jest.fn().mockResolvedValue(mockSubAccountsResponse),
    };

    const getSubAccounts = makeGetSubAccounts({
      accountId: 'mock_id',
      authToken: '123',
      service: mockSubAccountsService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: components,
      template: () => <payouts-list-core getPayouts={getPayouts} getSubAccounts={getSubAccounts} columns={defaultColumnsKeys} />,
    });

    await page.waitForChanges();

    const rows = page.root.querySelectorAll('[data-test-id="table-row"]');

    expect(rows.length).toBe(5);
    expect(page.root).toMatchSnapshot();
  });

  it('displays an error state on failed data fetch', async () => {
    const mockPayoutService = {
      fetchPayouts: jest.fn().mockRejectedValue(new Error('Fetch error'))
    };

    const getPayouts = makeGetPayouts({
      id: 'some-id',
      authToken: 'some-auth-token',
      service: mockPayoutService
    });

    const mockSubAccountsService = {
      fetchSubAccounts: jest.fn().mockResolvedValue(mockSubAccountsResponse),
    };

    const getSubAccounts = makeGetSubAccounts({
      accountId: 'mock_id',
      authToken: '123',
      service: mockSubAccountsService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: components,
      template: () => <payouts-list-core getPayouts={getPayouts} getSubAccounts={getSubAccounts} columns={defaultColumnsKeys} />,
    });

    await page.waitForChanges();

    const error = page.root.querySelector('[data-test-id="table-error-state"]');
    expect(error).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('emits an error event on failed data fetch', async () => {
    const mockPayoutService = {
      fetchPayouts: jest.fn().mockRejectedValue(new Error('Fetch error'))
    };

    const getPayouts = makeGetPayouts({
      id: 'some-id',
      authToken: 'some-auth-token',
      service: mockPayoutService
    });

    const mockSubAccountsService = {
      fetchSubAccounts: jest.fn().mockResolvedValue(mockSubAccountsResponse),
    };

    const getSubAccounts = makeGetSubAccounts({
      accountId: 'mock_id',
      authToken: '123',
      service: mockSubAccountsService,
      apiOrigin: 'http://localhost:3000'
    });

    const errorSpy = jest.fn();

    const page = await newSpecPage({
      components: components,
      template: () => <payouts-list-core getPayouts={getPayouts} getSubAccounts={getSubAccounts} columns={defaultColumnsKeys} onError-event={errorSpy} />,
    });

    await page.waitForChanges();

    expect(errorSpy).toHaveBeenCalledWith(expect.objectContaining({
      detail: {
        errorCode: 'fetch-error',
        message: 'Fetch error',
        severity: 'error',
      }
    }));
  });

  it('emits payout-row-clicked event on row click', async () => {
    const mockPayoutService = {
      fetchPayouts: jest.fn().mockResolvedValue(mockPayoutsResponse),
    };

    const getPayouts = makeGetPayouts({
      id: '123',
      authToken: '123',
      service: mockPayoutService
    });

    const mockSubAccountsService = {
      fetchSubAccounts: jest.fn().mockResolvedValue(mockSubAccountsResponse),
    };

    const getSubAccounts = makeGetSubAccounts({
      accountId: 'mock_id',
      authToken: '123',
      service: mockSubAccountsService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: components,
      template: () => <payouts-list-core getPayouts={getPayouts} getSubAccounts={getSubAccounts} columns={defaultColumnsKeys} />,
    });

    await page.waitForChanges();

    const rowClickedHandler = jest.fn();
    page.win.addEventListener('payout-row-clicked', rowClickedHandler);

    const rows = page.root.querySelectorAll('[data-test-id="table-row"]');

    if (rows.length > 0) {
      (rows[0] as HTMLElement).click();
      await page.waitForChanges();
    }
  });

  it('updates params and refetches data on pagination interaction', async () => {
    const mockPayoutService = {
      fetchPayouts: jest.fn().mockResolvedValue(mockPayoutsResponse),
    };

    const getPayouts = makeGetPayouts({
      id: '123',
      authToken: '123',
      service: mockPayoutService
    });

    const mockSubAccountsService = {
      fetchSubAccounts: jest.fn().mockResolvedValue(mockSubAccountsResponse),
    };

    const getSubAccounts = makeGetSubAccounts({
      accountId: 'mock_id',
      authToken: '123',
      service: mockSubAccountsService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: components,
      template: () => <payouts-list-core getPayouts={getPayouts} getSubAccounts={getSubAccounts} columns={defaultColumnsKeys} />,
    });
    await page.waitForChanges();

    page.rootInstance.handleClickNext('nextCursor');
    await page.waitForChanges();

    // The mock function should be called 2 times: once for the initial load and later after the pagination interaction
    expect(mockPayoutService.fetchPayouts).toHaveBeenCalledTimes(2);
    const updatedParams = page.rootInstance.params;
    expect(updatedParams.after_cursor).toBe('nextCursor');
  });
});
