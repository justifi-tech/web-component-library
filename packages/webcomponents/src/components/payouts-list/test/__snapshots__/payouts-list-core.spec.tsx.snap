// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`payouts-list-core displays an error state on failed data fetch 1`] = `
<payouts-list-core>
  <div class="table-wrapper" part="table-wrapper">
    <table class="table table-hover" part="table text color font-family background-color">
      <thead class="sticky-top table-head" part="table-head">
        <tr class="table-light text-nowrap" part="table-head-row">
          <th part="table-head-cell table-cell text color font-family background-color" scope="col" title="The date the payout was deposited">
            Created
          </th>
          <th part="table-head-cell table-cell text color font-family background-color" scope="col" title="The total amount of the payout">
            Amount
          </th>
          <th part="table-head-cell table-cell text color font-family background-color" scope="col" title="The real-time status of each payout">
            Status
          </th>
          <th part="table-head-cell table-cell text color font-family background-color" scope="col" title="The total payments amount associated with the payout">
            Payments
          </th>
          <th part="table-head-cell table-cell text color font-family background-color" scope="col" title="The total refunds amount associated with the payout">
            Refunds
          </th>
          <th part="table-head-cell table-cell text color font-family background-color" scope="col" title="The total fees amount associated with the payout">
            Fees
          </th>
          <th part="table-head-cell table-cell text color font-family background-color" scope="col" title="The total other amount">
            Other
          </th>
          <th part="table-head-cell table-cell text color font-family background-color" scope="col" title="Export CSV"></th>
        </tr>
      </thead>
      <tbody class="table-body">
        <tr>
          <td class="error-state" colspan="8" data-test-id="table-error-state" part="table-error text color font-family background-color" style="text-align: center;">
            An unexpected error occurred: Fetch error
          </td>
        </tr>
      </tbody>
      <tfoot class="sticky-bottom">
        <tr class="align-middle table-light">
          <td colspan="8" part="table-cell text color font-family background-color">
            <pagination-menu>
              <nav aria-label="Table pagination" class="d-flex gap-3 justify-content-end">
                <ul class="m-0 pagination">
                  <li class="disabled page-item">
                    <button class="page-link" part="button-disabled button text color font-family">
                      <span class="me-1" part="pagination-item-icon-previous pagination-item-icon text color font-family">
                        «
                      </span>
                      <span part="pagination-item-text text color font-family">
                        Previous
                      </span>
                    </button>
                  </li>
                  <li class="disabled page-item">
                    <button class="page-link" part="button-disabled button text color font-family">
                      <span part="pagination-item-text text color font-family">
                        Next
                      </span>
                      <span class="ms-1" part="pagination-item-icon-next pagination-item-icon text color font-family">
                        »
                      </span>
                    </button>
                  </li>
                </ul>
              </nav>
            </pagination-menu>
          </td>
        </tr>
      </tfoot>
    </table>
  </div>
</payouts-list-core>
`;

exports[`payouts-list-core renders properly 1`] = `
<payouts-list-core>
  <div class="table-wrapper" part="table-wrapper">
    <table class="table table-hover" part="table text color font-family background-color">
      <thead class="sticky-top table-head" part="table-head">
        <tr class="table-light text-nowrap" part="table-head-row">
          <th part="table-head-cell table-cell text color font-family background-color" scope="col" title="The date the payout was deposited">
            Created
          </th>
          <th part="table-head-cell table-cell text color font-family background-color" scope="col" title="The total amount of the payout">
            Amount
          </th>
          <th part="table-head-cell table-cell text color font-family background-color" scope="col" title="The real-time status of each payout">
            Status
          </th>
          <th part="table-head-cell table-cell text color font-family background-color" scope="col" title="The total payments amount associated with the payout">
            Payments
          </th>
          <th part="table-head-cell table-cell text color font-family background-color" scope="col" title="The total refunds amount associated with the payout">
            Refunds
          </th>
          <th part="table-head-cell table-cell text color font-family background-color" scope="col" title="The total fees amount associated with the payout">
            Fees
          </th>
          <th part="table-head-cell table-cell text color font-family background-color" scope="col" title="The total other amount">
            Other
          </th>
          <th part="table-head-cell table-cell text color font-family background-color" scope="col" title="Export CSV"></th>
        </tr>
      </thead>
      <tbody class="table-body">
        <tr data-row-entity-id="po_1a2b3c4d5e6f7g8h9i0j" data-test-id="table-row">
          <td part="table-cell-odd table-cell text color font-family background-color">
            <div class="fw-bold">
              Jan 18, 2024
            </div>
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            $226.02 USD
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            <span class="badge bg-success" part="badge font-family badge-success" title="Successfully deposited into your bank account">
              Paid
            </span>
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            $234.56 USD
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            $0.00 USD
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            $8.54 USD
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            $0.00 USD
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            <a href="#">
              CSV
            </a>
          </td>
        </tr>
        <tr data-row-entity-id="po_2b3c4d5e6f7g8h9i0j1a" data-test-id="table-row">
          <td part="table-cell-even table-cell text color font-family background-color">
            <div class="fw-bold">
              Jan 17, 2024
            </div>
          </td>
          <td part="table-cell-even table-cell text color font-family background-color">
            $226.02 USD
          </td>
          <td part="table-cell-even table-cell text color font-family background-color">
            <span class="badge bg-secondary" part="badge font-family badge-secondary" title="This payout initially failed; the funds have been forwarded to your next successful payout">
              Forwarded
            </span>
          </td>
          <td part="table-cell-even table-cell text color font-family background-color">
            $234.56 USD
          </td>
          <td part="table-cell-even table-cell text color font-family background-color">
            $0.00 USD
          </td>
          <td part="table-cell-even table-cell text color font-family background-color">
            $8.54 USD
          </td>
          <td part="table-cell-even table-cell text color font-family background-color">
            $0.00 USD
          </td>
          <td part="table-cell-even table-cell text color font-family background-color">
            <a href="#">
              CSV
            </a>
          </td>
        </tr>
        <tr data-row-entity-id="po_3c4d5e6f7g8h9i0j1a2b" data-test-id="table-row">
          <td part="table-cell-odd table-cell text color font-family background-color">
            <div class="fw-bold">
              Jan 10, 2024
            </div>
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            $226.02 USD
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            <span class="badge bg-primary" part="badge font-family badge-primary" title="Transfer to your bank account has been initiated">
              In Transit
            </span>
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            $234.56 USD
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            $0.00 USD
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            $8.54 USD
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            $0.00 USD
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            <a href="#">
              CSV
            </a>
          </td>
        </tr>
        <tr data-row-entity-id="po_4d5e6f7g8h9i0j1a2b3c" data-test-id="table-row">
          <td part="table-cell-even table-cell text color font-family background-color">
            <div class="fw-bold">
              Dec 28, 2023
            </div>
          </td>
          <td part="table-cell-even table-cell text color font-family background-color">
            $220.15 USD
          </td>
          <td part="table-cell-even table-cell text color font-family background-color">
            <span class="badge bg-primary" part="badge font-family badge-primary" title="Batched and scheduled to be transferred">
              Scheduled
            </span>
          </td>
          <td part="table-cell-even table-cell text color font-family background-color">
            $245.00 USD
          </td>
          <td part="table-cell-even table-cell text color font-family background-color">
            -$10.00 USD
          </td>
          <td part="table-cell-even table-cell text color font-family background-color">
            $14.85 USD
          </td>
          <td part="table-cell-even table-cell text color font-family background-color">
            $0.00 USD
          </td>
          <td part="table-cell-even table-cell text color font-family background-color">
            <a href="#">
              CSV
            </a>
          </td>
        </tr>
        <tr data-row-entity-id="po_5e6f7g8h9i0j1a2b3c4d" data-test-id="table-row">
          <td part="table-cell-odd table-cell text color font-family background-color">
            <div class="fw-bold">
              Dec 28, 2023
            </div>
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            $220.15 USD
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            <span class="badge bg-danger" part="badge font-family badge-danger" title="Transfer to your bank account failed">
              Canceled
            </span>
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            $245.00 USD
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            -$10.00 USD
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            $14.85 USD
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            $0.00 USD
          </td>
          <td part="table-cell-odd table-cell text color font-family background-color">
            <a href="#">
              CSV
            </a>
          </td>
        </tr>
      </tbody>
      <tfoot class="sticky-bottom">
        <tr class="align-middle table-light">
          <td colspan="8" part="table-cell text color font-family background-color">
            <pagination-menu>
              <nav aria-label="Table pagination" class="d-flex gap-3 justify-content-end">
                <ul class="m-0 pagination">
                  <li class="disabled page-item">
                    <button class="page-link" part="button-disabled button text color font-family">
                      <span class="me-1" part="pagination-item-icon-previous pagination-item-icon text color font-family">
                        «
                      </span>
                      <span part="pagination-item-text text color font-family">
                        Previous
                      </span>
                    </button>
                  </li>
                  <li class="disabled page-item">
                    <button class="page-link" part="button-disabled button text color font-family">
                      <span part="pagination-item-text text color font-family">
                        Next
                      </span>
                      <span class="ms-1" part="pagination-item-icon-next pagination-item-icon text color font-family">
                        »
                      </span>
                    </button>
                  </li>
                </ul>
              </nav>
            </pagination-menu>
          </td>
        </tr>
      </tfoot>
    </table>
  </div>
</payouts-list-core>
`;
